쥬스 메이커 [STEP 1] Grumpy, Cathy

grumpy-sw opened the pull request:
안녕하세요! @hyunable 

STEP1 PR 올립니다.🤗

---
고민했던 부분
1. JuiceMaker의 타입이 구조체이고 FruitStore의 타입이 클래스인 이유
2. 함수의 Naming이 적절한지
3. Juice의 recipe의 반환 타입을 튜플로 할 것인지, 딕셔너리로 할 것인지
4. FruitStore 생성 시 과일 딕셔너리 fruits의 초기화 방법
5. 재고 부족 외에 처리할 에러가 없는지
6. 열거형을 정의할 위치(사용자 타입 내부/외부)

---

피드백 받고싶은 부분
1. struct와 class의 사용
JuiceMaker는 struct로, FruitStore는 class를 사용하고 있어서 그 이유에 대해 생각해보았습니다. FruitStore에서는 fruits 딕셔너리의 value값이 계속 업데이트 되어야하므로 class를 사용하고, JuiceMaker의 경우엔 원본 값 변경을 하지 않기 때문에 굳이 class를 사용하지 않고
struct를 사용한 것이라 생각했습니다. 맞게 생각한 것인지 여쭤보고 싶고, class와 struct를 사용하는 기준이 궁금합니다.
2. class에서 init()사용
과일의 개수를 초기화하는 방법으로 딕셔너리를 선언하면서 초기화하는 방법과 생성자를 통해 초기화하는 방법 두 가지를 생각했습니다.
```swift
// 딕셔너리 선언과 함께 초기화
var fruits: [Fruit : Int] = [.strawberry : defaultStock,
                             .banana : defaultStock,
                             .pineapple : defaultStock,
                             .kiwi : defaultStock,
                             .mango : defaultStock]
```

```swift
// 생성자를 통한 초기화
init() {
    for fruit in Fruit.allCases {
        fruits[fruit] = defaultStock
    }
}
```
FruitStore class에서 fruits 딕셔너리의 값을 모두 10으로 설정해주는 방법으로 init()함수를 사용하였습니다. 한번에 값을 설정해줄 수 있다는 점과 추후에 case가 추가되었을 때 변경이 편리하다고 생각하여 변수에 직접 설정하지 않고 init()을 사용하였는데, 올바르게 생각한 것이 맞는지, 혹은 더 효율적으로 처리할 수 있는 방법이 있는지 hyunable의 의견을 듣고 싶어요!

3. dictionary와 tuple
JuiceMaker에서 JuiceRecipe를 딕셔너리로 설정해놓아서, hasEnoughFruitStock메서드에서 keys를 사용하기 위해 guard문을 세 번 사용하고 있습니다. tuple을 사용하면 아래와 같이 guard문을 한 번만 사용하게 되는데 어떤 방법이 더 효율적인지 궁금합니다.
```swift
func hasEnoughFruitStock(juice: Juice) -> Bool {
    let recipe: [(Fruit, Int)] = Juice.recipe(Juice.strawberryBananaJuice)()  // [(딸기, 10), (바나나, 1)]
    for ingredient in recipe {
        guard let currentStock = fruitStore.fruits[ingredient.0] else { return false }
        if currentStock < ingredient.1 {
            return false
        }
    }
    return true
}
```

4. 열거형 타입의 정의와 접근 방법에 대해서
처음에는 사용자 타입 바깥에 열거형을 정의했다가 과일의 종류에 해당하는 Fruit를 FruitStore 내부에, 쥬스의 종류에 해당하는 Juice를 JuiceMaker 내부로 이동했습니다. Juice의 경우 JuiceMaker 내에서만 사용하기 때문에 문제가 되지 않지만, Fruit의 경우 FruitStore와 JuiceMaker 모두에서 사용하고 있어 호출 방법이 타입을 거쳐 호출하도록 살짝 변경이 필요했습니다.
이 때 <u>사용자 타입 외부에</u> 열거형을 정의하여 모든 곳에서 접근이 가능하게 하는 방법과 <u>타입 내부</u>에 작성하여 ```FruitStore.Fruit``` 또는 ```JuiceMaker.Juice```처럼 접근하도록 하는 방법이 어떤 차이가 있는지 궁금합니다.

---

해결되지 않은 부분
~~1. "JuiceMaker는 FruitStore를 소유하고 있습니다." 라는 문구가 있는데요, 이 때문에 JuiceMakera 구조체에서 Fruit를 사용해야할 때 다음과 같이 사용해야 한다고 생각했습니다.~~
```swift
var fruitStore: FruitStore = FruitStore()
typealias JuiceRecipe = [[fruitStore.Fruit:Int]]
```
~~그러나 실제 코드 작성 결과 ```FruitStore.Fruit``` 와 같이 사용해야 했습니다. 위처럼 사용할 수 없는 이유가 무엇인지 알 수 있을까요?~~

위 부분은 중첩 타입에 대해 제가 잘못 이해했던 것 같습니다! 🙇🙇 공식 문서에 정의된 문맥 바깥에서 중첩 타입을 참조하려면 이름 앞에 중첩된 타입의 이름을 붙인다. 라고 언급되어 있네요..ㅎㅎ

Commit:
Feat: FruitStore 구현

FruitStore가 관리하는 과일 종류 설정, 과일의 초기 재고값 설정, 과일 수량 변경하는 메서드 구현

grumpy-sw committed:
feat: JuiceMaker 구현

Juice타입 열거형 추가, 재고 확인 함수 구현

Commit:
feat: 에러 처리 구현

enum사용하여 JuiceError 구현, consumeFruitStock 함수 추가

grumpy-sw committed:
refactor: 소스 리팩토링

JuiceMaker에서 Juice.strawberryBananaJuice로 입력되어 있는 것 수정.
열거형 타입에서 Juice. -> .으로 수정

Commit:
refactor: typealias 생성, 파라미터 변경

JuiceRecipe라는 이름의 typealias 생성, 함수의 파라미터를 juiceRecipe 타입으로 변경

grumpy-sw committed:
refactor: Fruit 열거형 FruitStore 아래로 이동

Commit:
docs: 쥬스메이커 readme.md


readme 작성

hyunable commented on JuiceMaker/JuiceMaker/Model/FruitStore.swift:
```diff
@@ -1,12 +1,29 @@
+        case strawberry, banana, pineapple, kiwi, mango
+    }
+    
+    let defaultStock = 10
```
`수량`이라는 점을 알 수 있는 좀 더 디테일한 변수명이면 좋을 것 같아요 ㅎㅎ

```suggestion
    let numberOfStock = 10
```

```suggestion
    let defaultNumberOfStock = 10
```

위와 같은 변수명들은 어떨까요?

grumpy-sw replied:
변경했습니다!

hyunable commented on JuiceMaker/JuiceMaker/Model/FruitStore.swift:
```diff
@@ -1,12 +1,29 @@
+
 // 과일 저장소 타입
 class FruitStore {
+    enum Fruit: CaseIterable {
```
enum에서는 보통 `type`이라는 단어를 많이 붙여서 네이밍해주는 편인것 같아요
ex, ErrorType, AuthorType ...

요것도 `Fruit` 타입명만 보면 과일 class로 느껴질 수 있을 것 같아요 ㅎㅎ type을 붙여보는건 어떨까요?

cathy171 replied:
FruitType으로 변경하였습니다.

hyunable commented on JuiceMaker/JuiceMaker/Model/JuiceMaker.swift:
```diff
@@ -1,12 +1,59 @@
 // 쥬스 메이커 타입
 struct JuiceMaker {
+    var fruitStore: FruitStore = FruitStore()
+    typealias JuiceRecipe = [[FruitStore.Fruit:Int]]
```
🙋🏻‍♀️ typealias 는 보통 가장 최상단에 기재해주기는 합니다. 스타일이기때문에 참고만해주세용 ㅎㅎ

cathy171 replied:
typealias를 가장 윗줄로 변경하였습니다.

hyunable commented on JuiceMaker/JuiceMaker/Model/JuiceMaker.swift:
```diff
@@ -1,12 +1,59 @@
 
 // 쥬스 메이커 타입
 struct JuiceMaker {
+    var fruitStore: FruitStore = FruitStore()
```
fruitStore가 다른 store로 대체되거나 변경되지 않기 때문에 `let`으로 써주면 좋을 것 같아요

hyunable replied:
외부에서 fruitStore의 재고를 막 변경할 수도 있을 것 같아요  👀 fruitStore는 내부에서만 쓰이고있고 외부에서는 몰라도 되는 프로퍼티기 때문에 private로 접근제어자를 변경해주는건 어떨까요?

cathy171 replied:
fruitStore를 프로퍼티를 private let으로 변경하였습니다.

hyunable commented on JuiceMaker/JuiceMaker/Model/JuiceMaker.swift:
```diff
@@ -1,12 +1,59 @@
+            case .mangoJuice:
+                return [[.mango : 3]]
+            case .mangoKiwiJuice:
+                return [[.mango : 2], [.kiwi : 1]]
```
Dictionary는 책과 같아서 다양한 키값에 따른 밸류를 담을 수 있어요~!

```suggestion
                return [.strawberry: 16]
            case .bananaJuice:
                return [.banana: 2]
            case .kiwiJuice:
                return [.kiwi: 3]
            case .pineappleJuice:
                return [.pineapple: 2]
            case .strawberryBananaJuice:
                return [.strawberry: 10, .banana: 1]
            case .mangoJuice:
                return [.mango: 3]
            case .mangoKiwiJuice:
                return [.mango: 2, .kiwi: 1]
```

혹시 위와 같이 변경해주는 건 어떨가요? 
혹시 dictionary를 담은 array 형태로 레시피를 구현한 사유가 따로 있다면 설명이 듣고싶습니다 :)


> ⚠️ Lint: 보통 키값과 콜론 사이에는 공백을 없애고 밸류와 콜론 사이에는 공백을 넣어주세요 ex, ["hyunah": 999]

cathy171 replied:
array에 따로 담지 않고 dictionary에 담는 방식으로 변경하였고, 공백도 수정하였습니다!

grumpy-sw replied:
키를 여러 개 갖는 경우 딕셔너리가 여러 개가 되는 건줄 알았는데 이렇게 사용할 수 있었군요!

hyunable commented on JuiceMaker/JuiceMaker/Model/JuiceMaker.swift:
```diff
@@ -1,12 +1,59 @@
+        }
+    }
+    
+    // 주스를 만드는 함수
```
👍 상세한 주석 너무 좋아요~!
하지만 개인적으로는 주석이 없이 네이밍으로도 충분히 역할을 유추할 수 있다면 주석은 생략해도 좋을 것 같아요 ㅎㅎ

grumpy-sw replied:
짝 프로그래밍을 하면서 설명을 위해 만들었던 주석인데
깜빡하고 안지웠군요.. 😅
수정했습니다!

hyunable commented on JuiceMaker/JuiceMaker/Model/JuiceMaker.swift:
```diff
@@ -1,12 +1,59 @@
+    }
+    
+    // 재고가 충분한지 확인하는 함수
+    func hasEnoughFruitStock(recipe: JuiceRecipe) -> Bool {
```
해당 함수는 makeJuice내부에서만 사용되고 있는데요 ㅎㅎ
외부에 노출될 필요 없는 메서드나 프로퍼티는 기본적으로 모두 private로 접근제어를 설정해주는 것이 좋아요~!

grumpy-sw replied:
현재 시점에서는 외부에서 사용할 것으로 예상되는 함수는 makeJuice뿐이네요.
이 함수를 제외하고 나머지는 private으로 변경했습니다!

hyunable commented on JuiceMaker/JuiceMaker/Model/JuiceMaker.swift:
```diff
@@ -1,12 +1,59 @@
+    func consumeFruitStock(recipe: JuiceRecipe) {
+        for ingredient in recipe {
+            guard let fruit = ingredient.keys.first else { return }
+            guard let amount = ingredient[fruit] else { return }
```
요렇게 fruite와 amount값을 빼오는 로직이 hasEnoughFruitStock에서도 이루어지고 consumeFruitStock에서도 이루어지고있어요 ㅎㅎ recipe에 있는 fruite의 양이 많을 경우 굉장히 많은 리소스가 낭비될 수 있을 것 같아요~! 

요 로직을 합칠 수 있는 방법은 없을까요? 과일을 만드는과정에서 한번의 체크로도 가능할 것 같아요

grumpy-sw replied:
recipe가 필요로 하는 과일의 종류가 여러 종류인 경우, 재고 확인 직후에 바로 재고 소모를 할 수 없는 상황이기 때문에 
필요한 과일의 종류와 개수를 확인하는 과정을 두 번씩 거치고 있습니다.
예를 들어 '딸바쥬스'가 주문이 들어왔다면, 딸기와 바나나가 각각 10개, 1개 이상 있는지 확인하고 재고가 충분하다면 consumeFruitStock을 호출하여 필요한 과일을 소모하는 방식으로 동작하는 것이죠.

저희가 생각하기에 한번의 체크로 가능하도록 코드를 변경한다면, recipe의 element 수만큼 반복문을 돌면서 재고를 확인 및 소모하고 부족한 재고가 있을 시 원래대로 롤백하는 함수가 추가되어야 할 것 같은데, 오히려 함수를 복잡하게 만들 것 같다는 생각이 들었습니다. 이 부분에 대한 조언을 부탁드립니다 🙏🙏

hyunable replied:
제생각에 precondition check 없이 레시피대로 재고를 소진하는과정에서 "음 사과2개 망고1개가 필요하네" , " 엇? 근데 사과 재고가 1개야", "에러를 던져줘" 와 같이 재료를 소진하는 과정에서 에러를 던질수있지않을까?
그럼 재고체크와 레시피를 소진할때 매번 값을 가져올 필요는 없을 것 같아서요 ㅎㅎ
혹시 이해안가시는 부분이나 이해가 다른 부분이 있다면 디스코드로 말씀해주세용~!

grumpy-sw replied:
음 제가 hyunable의 코멘트를 완벽하게 이해한 것 같지는 않지만 일단 소스를 수정해보았습니다. 
재고가 부족하면 에러를 던지고, 재고가 부족하지 않다면 과일의 요구량 배열을 만들어서 여기에 추가한 뒤 배열의 크기만큼 decreaseStock을 호출하도록 구성했습니다. 다만 말씀하셨던 두 로직을 합치는 해답인 것 같지는 않아 보입니다. 
레시피에서 필요한 과일의 종류와 양을 확인하고, 현재 재고량과 비교하는 과정에서 각 하나씩 접근하는 반복 로직이 필연적으로 사용될 것 같은데, 비교 혹은 재고 소진을 한번에 할 수 있는 방법이 있는지는 잘 떠오르지 않네요...ㅠㅠ

hyunable replied:
아핫 ㅎㅎ 재고가 부족하면 에러를 던지도록 로직을 변경해주셨는데요, 그렇다면 주스를 만들기전에 미리 재고비교를 하지 않아도 될 것 같아요~! 지금 코드를 보면 두군데에서 같은 Error를 던지고있어용 (decreaseStock과 비교구문) 
비교구문을빼보는건 어떨까요?

hyunable commented on JuiceMaker/JuiceMaker/Model/FruitStore.swift:
```diff
@@ -1,12 +1,29 @@
+    
+    func decreaseStock(fruit: Fruit, amount: Int) {
+        guard let currentStock = fruits[fruit] else { return }
+        fruits.updateValue(currentStock - amount, forKey: fruit)
```
재고수량 이하로 decreaseStock에서 amount값이 들어오게된다면 재고는 마이너스값을 가질수도 있을 것 같아요 🤔
위와 같은 케이스도 대응을 해주면 좋을 것 같습니다 :)

grumpy-sw replied:
```decreaseStock```을 호출하기 전에 조건으로 걸러내기 때문에 별도의 예외 처리를 하지 않았었습니다.. decreaseStock 함수 내에서 currentStock과 amount를 비교한 뒤 currentStock < amount인 경우 지정해 둔 Error인 ```notEnoughStock```을 throw하게 만들었습니다~!

```swift
func decreaseStock(fruit: FruitType, amount: Int) throws {
    ...
    guard currentStock >= amount else { throw ErrorType.notEnoughStock }
    fruits.updateValue(currentStock - amount, forKey: fruit)
}

```

hyunable replied:
decreaseStock 를 호출 전에 체크를 해주는것 너무 좋습니다~! 하지만 협업하는 동료가 해당 사항을 놓칠수도 있고, precondition 체크를 해주지 않는다면 오류가발생하기때문에 decreaseStock 내부에서 처리해주면 좋지 않을까라는 관점에서 말씀드려봤습니다 ㅎㅎ

hyunable commented on JuiceMaker/JuiceMaker/Model/JuiceMaker.swift:
```diff
@@ -1,12 +1,59 @@
+                return false
+            }
+        }
+        return true
```
조건문이 함께 잇다면 true false를 직접 return 해주기보다
조건문의 결과로 return해주는 것이 구문도 줄이고 직관적일 수있어요 ㅎㅎ

`return currentStock >= amount` 요고로 대체할 수 있을 것 같습니다

grumpy-sw replied:
위에서 언급한 것처럼 여러 종류의 과일이 필요한 경우 모두 체크하기 위해 리턴문을 나누었습니다. ```return currentStock >= amount```로 사용한다면 첫 번째 재료가 ```true```조건일 때 다음 과일을 체크하지 않고 그대로 함수의 결과를 ```true```로 반환할 것으로 예상됩니다. 그래서 이 부분은 수정하지 않고 그대로 두었는데, 혹시 구문도 직관적으로 변경하면서 로직도 보다 간단하게 수정할 방법이 있을까요?😱

hyunable replied:
```swift
if currentStock < amount {
   return false
} else { 
   return true 
}
```

아핫 제가 마지막 코드 부분을 위와같이 보았네요 ㅎㅎ
네넵 지금 그대로 유지해도 괜찮을 것 같습니다

hyunable reviewed (commented):
고생 많으셨어요 :)
코멘트 한번씩만 확인해주시고 반영하신부분은 리액션을, 궁금한점 있으면 코멘트에 댓글달아주세용~!

hyunable commented:
### 피드백 받고싶은 부분
1. struct와 class의 사용
JuiceMaker는 struct로, FruitStore는 class를 사용하고 있어서 그 이유에 대해 생각해보았습니다. FruitStore에서는 fruits 딕셔너리의 value값이 계속 업데이트 되어야하므로 class를 사용하고, JuiceMaker의 경우엔 원본 값 변경을 하지 않기 때문에 굳이 class를 사용하지 않고
struct를 사용한 것이라 생각했습니다. 맞게 생각한 것인지 여쭤보고 싶고, class와 struct를 사용하는 기준이 궁금합니다.
> 고민을 잘 해주신것 같아요 :) 말씀하신대로 FruitStore의 경우 데이터를 갖고있고 데이터를 다루고있기때문에 class에 적합하다고 보여요. 이 외에도 class와 struct를 쓰는 기준에 대해서 고민이 많으실텐데요 ㅎㅎ 간단하게 `틀`같은 역활을 하고있다면 struct를 데이터를 다루고 해당 객체에 접근해서 데이터를 변경하고 그 변경된 값이 반영이 되어야 한다면 (값으로 복사되는 것이 아닌) class로 써주는 것이 적합하다고 생각해주시면 됩니다 

2. class에서 init()사용
과일의 개수를 초기화하는 방법으로 딕셔너리를 선언하면서 초기화하는 방법과 생성자를 통해 초기화하는 방법 두 가지를 생각했습니다.
```swift
// 딕셔너리 선언과 함께 초기화
var fruits: [Fruit : Int] = [.strawberry : defaultStock,
                             .banana : defaultStock,
                             .pineapple : defaultStock,
                             .kiwi : defaultStock,
                             .mango : defaultStock]
```

```swift
// 생성자를 통한 초기화
init() {
    for fruit in Fruit.allCases {
        fruits[fruit] = defaultStock
    }
}
```
FruitStore class에서 fruits 딕셔너리의 값을 모두 10으로 설정해주는 방법으로 init()함수를 사용하였습니다. 한번에 값을 설정해줄 수 있다는 점과 추후에 case가 추가되었을 때 변경이 편리하다고 생각하여 변수에 직접 설정하지 않고 init()을 사용하였는데, 올바르게 생각한 것이 맞는지, 혹은 더 효율적으로 처리할 수 있는 방법이 있는지 hyunable의 의견을 듣고 싶어요!

> 넵~! 작성해주신 초기화 방법 너무 좋다고 생각합니다 :)  말씀하신대로 케이스가 추가되었을때 다른 추가 코드 없이도 `각 과일은 10개만큼 재고를 가져야한다` 는 spec을 너무 잘 충족시켜줄수있어서입니다. 만약 각 과일이 FruiteStore를 생성할때마다 과일 종류와 수량을 다르게 만들어주어야한다면 parameter를 만들고 입력받는 방식이 적합했을거에요 ㅎㅎ

3. dictionary와 tuple
JuiceMaker에서 JuiceRecipe를 딕셔너리로 설정해놓아서, hasEnoughFruitStock메서드에서 keys를 사용하기 위해 guard문을 세 번 사용하고 있습니다. tuple을 사용하면 아래와 같이 guard문을 한 번만 사용하게 되는데 어떤 방법이 더 효율적인지 궁금합니다.
```swift
func hasEnoughFruitStock(juice: Juice) -> Bool {
    let recipe: [(Fruit, Int)] = Juice.recipe(Juice.strawberryBananaJuice)()  // [(딸기, 10), (바나나, 1)]
    for ingredient in recipe {
        guard let currentStock = fruitStore.fruits[ingredient.0] else { return false }
        if currentStock < ingredient.1 {
            return false
        }
    }
    return true
}
```

> 코드리뷰에 기재해 놓은 것 처럼 dictionary 형태도 좋고, 말씀주신 튜플형태도 좋습니다. 다만 튜플을 쓴다면, 라벨을 같이 넣어주는게 좋을 것 같구요 (ex, (fruit: Fruit, amount: Int) 왜냐하면 사용처에서 0, 1과같은 숫자로 접근해서 쓰는 것은 비추드립니다 ㅠㅠ 직관적이지 못하기 때문이죠. dictionary로 쓰면 좋은 점은 FruitStore에서 [Fruit: Int] 형태로 재고를 관리하고있는데, typealias 선언 없이 해당 타입을 그대로 쓸 수 있다는 점 같아요 ㅎㅎ

4. 열거형 타입의 정의와 접근 방법에 대해서
처음에는 사용자 타입 바깥에 열거형을 정의했다가 과일의 종류에 해당하는 Fruit를 FruitStore 내부에, 쥬스의 종류에 해당하는 Juice를 JuiceMaker 내부로 이동했습니다. Juice의 경우 JuiceMaker 내에서만 사용하기 때문에 문제가 되지 않지만, Fruit의 경우 FruitStore와 JuiceMaker 모두에서 사용하고 있어 호출 방법이 타입을 거쳐 호출하도록 살짝 변경이 필요했습니다.
이 때 <u>사용자 타입 외부에</u> 열거형을 정의하여 모든 곳에서 접근이 가능하게 하는 방법과 <u>타입 내부</u>에 작성하여 ```FruitStore.Fruit``` 또는 ```JuiceMaker.Juice```처럼 접근하도록 하는 방법이 어떤 차이가 있는지 궁금합니다.

> 너무 좋은 고민 포인트인거같아요 :) `FruitStore.Fruit` 로 쓰거나, 해당 enum을 외부로 옮겨서 `Fruit`으로 쓰거나 퍼포먼스 상으로는 문제가 없습니다. 다만 해당 enum이 외부에서도 쓰고있는지와 어떤 클래스와 디펜던시가 있는지 두가지 포인트에 대해서 고민해서 결정내리면 좋을 것 같아요. 어떤 클래스에 종속적인 enum 이라면 ex, `WhiskeyStore` 클래스에서 위스키의 종류를 다루는 enum이 있고, 외부에서는 쓰이지 않고 있을때 WhiskeyStore 클래스에 종속적이라고 볼 수 있을 것 같아요. 그런 경우에는 내부에, 그렇지 않다면 외부에 선언해주시는 편이 좋습니다 ㅎㅎ
또하나 예를 들면,  BloodType이라는 enum이 있는데 `한국인` 클래스 내부에 있다면 ..? `미국인`, `일본인` 클래스에서도 쓰일 수있고 `한국인` 클래스에 종속적이지 않기 때문에 외부에 독립적으로 선언해주시는 것이 좋습니다 ㅎㅎ
그럼 Fruit 는 어디에 선언되는 것이 좋을까요?

Commit:
refactor: Step1 수정

변수 네이밍 변경, enum 네이밍 변경, 메서드와 프로퍼티를 private으로 변경, array 제거, 주석 제거

grumpy-sw committed:
fix: 재고가 음수가 될 수 있는 경우에 대한 예외처리 추가

JuiceError -> ErrorType으로 변경
decreaseStock에서 currentStock < amount일 경우 ErrorType.notEnoughStock throw

Commit:
refactor: consumeFruitStock을 private으로 접근 제어 설정, if문을 guard문으로 변경

grumpy-sw committed:
refactor: makeJuice 함수 분리

requiredNumberOfFruits, consumeFruitStock으로 분리

grumpy-sw commented:
@hyunable 

피드백 주신 부분 반영해서 소스 수정 후 커밋 했습니다. 감사합니다!

hyunable commented on JuiceMaker/JuiceMaker/Model/ErrorType.swift:
```diff
@@ -1,5 +1,5 @@
 import Foundation
 
-enum JuiceError: Error {
+enum ErrorType: Error {
```
Error type은 너무 제네럴해서 어떤에러를 담고있는지 추측하기가 힘든것 같아요 ㅎㅎ 🤔
JuiceMaker에서 다뤄지는 에러들이니 `JuiceMakerError` 는 어떨까용?  (에러타입을 다루는 enum은 기존 네이밍 그대로 ~~Error요렇게 네이밍을 일반적으로하는것같아요 ㅎㅎ)

Commit:
refactor: JuiceMaker 재고 소모하는 함수 로직 변경, Error enum 네이밍 변경

grumpy-sw committed:
refactor: ErrorType 이름 변경, requiredNumberOfFruits 함수 제거

Commit:
feat: 과일 재고 체크하는 함수 생성

grumpy-sw committed:
feat: 쥬스 제조 함수에서 재고를 확인하는 함수 호출하도록 변경, 에러 타입 추가

haveEnoughFruitStock -> checkEnoughFruitStock 함수 이름 변경
makeJuice 함수 내에서 호출하도록 변경
재고가 0일 시 에러 유형인 outOfFruitStock 케이스를 JuiceMakerError에 추가

hyunable reviewed (approved):
수고많으셨습니다👍🏻
